server/channels/app/oembed/endpoint.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package oembed
+
+import (
+	"net/url"
+	"regexp"
+)
+
+//go:generate go run ./generator/providers_generator.go
+
+type ProviderEndpoint struct {
+	URL      string
+	Patterns []*regexp.Regexp
+}
+
+func (e *ProviderEndpoint) GetProviderURL(requestURL string) string {
+	// This error is checked when generating the list of providers
+	url, _ := url.Parse(e.URL)
+
+	query := url.Query()
+	query.Add("format", "json")
+	query.Add("url", requestURL)
+	url.RawQuery = query.Encode()
+
+	return url.String()
+}
+
+// FindEndpointForURL returns a ProviderEndpoint for a given URL if it matches one that's supported by us. Returns nil
+// if none of the supported providers match the given URL.
+func FindEndpointForURL(requestURL string) *ProviderEndpoint {
```

Comment:
Is this going to be a performance concern? 

- Should we add a benchmark?
- Should we consider caching?

It's fine if the answer is no to the above, having the single provider and the regexp being precompiled should help. What I am not sure about is how frequently we'd be calling this from the app side.

Reply:
No, I don't expect it to be a performance concern. This is called by `App.getLinkMetadata` whenever the server sends posts to the client, but because `getLinkMetadata` caches its results, this should only ever be called once per post (outside of very rare cases but that code is idempotent so that's acceptable). Even when this is called, it's followed by an HTTP request to a third party server regardless of the result, and that'd have a relatively much larger effect.

To answer your questions more directly, I'd say no to benchmarks and it's basically already cached